datashader/composite.py

```diff
@@ -109,3 +110,45 @@ def saturate(src, dst):
     g = (factor * sg + dg * da)/a
     b = (factor * sb + db * da)/a
     return combine_scaled(r, g, b, a)
+
+
+
+def arr_operator(f):
+    """Define and register a new image composite operator"""
+
+    if jit_enabled:
+        f2 = nb.vectorize(f)
+        f2._compile_for_argtys(
+           (nb.types.int32, nb.types.int32))
+        f2._compile_for_argtys(
+           (nb.types.int64, nb.types.int64))
+        f2._compile_for_argtys(
+            (nb.types.float32, nb.types.float32))
+        f2._compile_for_argtys(
+            (nb.types.float64, nb.types.float64))
+        f2._frozen = True
+    else:
+        f2 = np.vectorize(f)
+
+    composite_op_lookup[f.__name__] = f2
+    return f2
+
+
+@arr_operator
+def source_arr(src, dst):
+    if src:
+        return src
+    else:
+        return dst
+
+@arr_operator
+def over_arr(src, dst):
+    return src + dst
+
+@arr_operator
+def add_arr(src, dst):
+    return src + dst
+
+@arr_operator
+def saturate_arr(src, dst):
+    return src + dst
```

Comment:
Seems like for arrays we also need max and min operators.  Maybe source, over, or saturate could be most appropriately interpreted as one or both of those?  If not, maybe need to add separate max and min operators, presumably both for images and for arrays.

Reply:
This is leading me to believe that maybe we do some validation about the operators used. We can default to `add` which makes sense for RGB and aggregate arrays and then complain if an unsuitable operator is chosen for the supplied data type. 

Then we could offer `min` and `max` operators for use with aggregate arrays only instead of trying to map these two concepts back to the set of operators that make sense with RGB.

Reply:
Is this resolved, then?

Reply:
I think the idea of 'saturate' being `src+dst` might be a bit iffy but the 'min' and 'max' operators are now offered and I think *those* semantics are clear at least. I suppose the issue is whether we want to support the same operator names as RGB? Having to support some operators for RGB only and a different set for arrays would be a pain so this kind of behavior isn't totally unreasonable imho as long as it is described in the docs.

Reply:
If we leave `saturate` and `over` defined for arrays, then we'd presumably not mention them in the docs, but I'm inclined to delete them if they don't make sense for arrays.  That would mean changing the default `how` to None, with code selecting `add` if it's an array and `over` if it's an Image, I guess?

Reply:
I don't think `how=None` is particularly clear as there *is* a default 'how' which (in my opinion) ought to be explicit. Tricky! Maybe we could do the following?

1) As we want to encourage array spreading instead of the questionable RGB spreading, I think the default should be driven by that choice. So the default should be `how='add'` but there is already an image operator named 'add'...
2) ... which means I could rename the array `add` operator to `'sum'`? I think 'source' is the only operator that really makes sense for both images and array cases.
3) So the default could be `how='sum'` in which case image inputs could warn the user that they are switching to `how='over'` for backwards compatibility?
4) Given a default that works for both images and arrays, I think we should then do some validation. Images can then only use `source`, `over`, `saturate` and `add` which arrays can only use `source`, `sum`, `min` and `max`

Would this make sense?

Reply:
I think that will be confusing, as there is no semantic difference between `add` and `sum`.  Warning that we'll switch over to `over` for backwards compatibility seems like a pain for users; there's nothing wrong with their current code, and they should expect it to keep working with no warnings unless we wished to warn about using RGB spreading at all (which even if we did wouldn't be for another 6 months at the earliest given how long RGB spreading has been the only spreading available!).  I do think we should do validation, but I find the issues with changing the default to None much less than the issues with these other alternatives.

Reply:
Ok...we can do that but we will need to state that the default 'how' is 'over' and 'add' for images and arrays respectively both in the docstrings and the online docs.

Reply:
Sure, but online docs might not even mention any details about RGB spreading, so it might not come up there at all...